### PR TITLE
Re cycle IP Address

### PR DESCRIPTION
**Pull request by 6mxo89sqh**
### This is POC/Education purpose. Use at your own risk.

Recycle IP Address to get more api call limit

Read the setup notes in [docs/AWS.md](https://github.com/6mxo89sqh/covid-vaccine-booking/blob/rate-limit-ip-restrictions/docs/AWS.md)

Please test and let me know your feedback.

#### If you have tested this PR and found it working fine or have any feedback/issues, please leave a comment.

### CODE REVIEW

**Comment by bombardier-gif** [comment]
Thanks @6mxo89sqh Do you think we are good to merge this to master?

**Comment by 6mxo89sqh** [comment]
@bombardier-gif, i have tested it from my side. but lets wait for atleast one to test this. It should be okay to merge as it doesn't disturb the non AWS users. But, i ll still prefer if we wait for someone to test it.

**Comment by 6mxo89sqh** [comment]
@shivaslifzz As you are using AWS, can you please test this branch?

**Comment by shivaslifzz** [comment]
@6mxo89sqh  I tested this code in AWS EC2(UBUNTU) instance with 2nd search option with refresh rate of 3 seconds I am getting the issue related to rate limit 

image

image

**Comment by 6mxo89sqh** [comment]
@shivaslifzz from your logs, I see that you haven't specified aws region. 

Please run the following command before running the script:
`export AWS_DEFAULT_REGION=ap-south-1`

I hope you have configure authentication as mentioned in doc.

**Comment by shivaslifzz** [comment]
> @shivaslifzz from your logs, I see that you haven't specified aws region.
> 
> Please run the following command before running the script:
> `export AWS_DEFAULT_REGION=ap-south-1`
> 
> I hope you have configure authentication as mentioned in doc.

ENI is chargeable bro

**Comment by 6mxo89sqh** [comment]
@shivaslifzz One EIP per ec2 instance isn't charged. Refer [here](https://aws.amazon.com/ec2/pricing/on-demand/). Assigning EIP for second ENI is not mandatory. If you assign EIP, you can directly login using to that box. Otherwise, you can ssh from another running machine in same network on private ip of secondary interface.

**Comment by shivaslifzz** [comment]
> @shivaslifzz One EIP per ec2 instance isn't charged. Refer [here](https://aws.amazon.com/ec2/pricing/on-demand/). Assigning EIP for second ENI is not mandatory. If you assign EIP, you can directly login using to that box. Otherwise, you can ssh from another running machine in same network on private ip of secondary interface.



> @shivaslifzz One EIP per ec2 instance isn't charged. Refer [here](https://aws.amazon.com/ec2/pricing/on-demand/). Assigning EIP for second ENI is not mandatory. If you assign EIP, you can directly login using to that box. Otherwise, you can ssh from another running machine in same network on private ip of secondary interface.

I tried it bro but no luck 

I followed below steps 
1) Created json file in IAM policy
2) Created Elastic Ip and attached it to EC2 instance
3) And connected to the Ec2 instance using EIP only but still getting rate limit issue

**Comment by 6mxo89sqh** [comment]
@shivaslifzz did you attach the policy to the instance? can you please share logs? (You can [connect with me](https://discord.gg/394ZXQeW) to debug.)

**Comment by shivaslifzz** [comment]
> @shivaslifzz did you attach the policy to the instance? can you please share logs? (You can [connect with me](https://discord.gg/394ZXQeW) to debug.)

At what time

**Comment by 6mxo89sqh** [comment]
@shivaslifzz I am online right now.

**Comment by shivaslifzz** [comment]
> @shivaslifzz I am online right now.

Lets meet now

**Comment by 6mxo89sqh** [comment]
@shivaslifzz , [here](https://discord.gg/394ZXQeW).

**Comment by shivaslifzz** [comment]
> @shivaslifzz did you attach the policy to the instance? can you please share logs?

@6mxo89sqh thanks man it's working fine after adding the iam role

**Comment by bombardier-gif** [comment]
@6mxo89sqh can you resolve the conflicts please, if it doesnt break non aws users, then i will merge this

**Comment by 6mxo89sqh** [comment]
> @6mxo89sqh can you resolve the conflicts please, if it doesnt break non aws users, then i will merge this

@bombardier-gif done!

**Comment by ketan** [comment]
Can confirm this works, however instructions are a bit unclear, so I got kicked out even though I was using the secondary IP. It's probably best if we're able to pass all information (including the kvdb key) over CLI, then we can just leave the docker  container running as a daemon.

**Comment by 6mxo89sqh** [comment]
> Can confirm this works, however instructions are a bit unclear, so I got kicked out even though I was using the secondary IP. It's probably best if we're able to pass all information (including the kvdb key) over CLI, then we can just leave the docker container running as a daemon.

Yes @ketan you can pass all info via cli and run it as daemon. Thats there in readme. 
for kvdb, u need to put `export KVDB_BUCKET=ur key` before you start the script. once the json is generated, you can run it as docker daemon (`--no-tty` flag to docker) :) 

I will add the doc on how to assign secondary EIP. Thanks for the feedback.

**Comment by 6mxo89sqh** [comment]
@ketan updated the doc.

@bombardier-gif Good to merge. Few users have tried it.

**Review by shailesh — COMMENTED** [review]
LGTM, let me test this today sometime
 I'll leave a comment if any

**Comment by AnjunaBoy** [comment]
Hi @6mxo89sqh @shailesh : a question from a noob here. So the changes being tested to cycle the IP, that feature can only be used by some one using AWS? Or will someone like me using Python also be able to use the feature? 

I have no knowledge of AWS EC2 but happy to learn in order to help folks around me. Am i right in assuming that if I dont use AWS, I cannot cycle my IP and will be rate limited after every 20 calendar searches?

**Comment by shivaslifzz** [comment]
> Hi @6mxo89sqh @shailesh : a question from a noob here. So the changes being tested to cycle the IP, that feature can only be used by some one using AWS? Or will someone like me using Python also be able to use the feature?
> 
> I have no knowledge of AWS EC2 but happy to learn in order to help folks around me. Am i right in assuming that if I dont use AWS, I cannot cycle my IP and will be rate limited after every 20 calendar searches?

You don't need aws you can follow the same old approach it will work

**Comment by 6mxo89sqh** [comment]
> Hi @6mxo89sqh @shailesh : a question from a noob here. So the changes being tested to cycle the IP, that feature can only be used by some one using AWS? Or will someone like me using Python also be able to use the feature?
> 
> I have no knowledge of AWS EC2 but happy to learn in order to help folks around me. Am i right in assuming that if I dont use AWS, I cannot cycle my IP and will be rate limited after every 20 calendar searches?

Please wait for PRs to be merged.

File: docs/AWS.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                "ec2:CreateTags",
+                "ec2:AssociateAddress",
+                "ec2:AllocateAddress"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+3. Create IAM Role by following [this guide](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#create-iam-role). For step number 4, of AWS guide, use the IAM Policy name just created in step number 2.
```
Comment (Nakul93) [inline docs/AWS.md:25]: Could you explain these steps in a much more clearer way. In step2 it says you need to paste the json for step no5 which is misleading

**Review by Nakul93 — CHANGES_REQUESTED** [review]

File: docs/AWS.md (hunk 1 above)
Comment (6mxo89sqh) [reply docs/AWS.md]: Done. Basically, it provides what to paste while following the aws guide.

**Review by Nakul93 — APPROVED** [review]

**Comment by 6mxo89sqh** [comment]
@bombardier-gif / @shailesh / @Nakul93 Should we merge this?

**Comment by AnjunaBoy** [comment]
> > Hi @6mxo89sqh @shailesh : a question from a noob here. So the changes being tested to cycle the IP, that feature can only be used by some one using AWS? Or will someone like me using Python also be able to use the feature?
> 
> > 
> 
> > I have no knowledge of AWS EC2 but happy to learn in order to help folks around me. Am i right in assuming that if I dont use AWS, I cannot cycle my IP and will be rate limited after every 20 calendar searches?
> 
> 
> 
> You don't need aws you can follow the same old approach it will work

Thanks for confirming. Looking forward to the merge.

**Pull request merged**